### PR TITLE
feat: migrate to the new version of Jupiter's SecurityPolicy

### PR DESCRIPTION
**Pull request by marcambier**
**Issue**

https://github.com/gravitee-io/issues/issues/7991

---
**Gravitee.io Automatic Deployment**

🚀 A prerelease version of this package has been published on Gravitee's private artifactory, you can:
 - use it directly by updating your project with version: `1.22.0-feat-jupiterPlans-SNAPSHOT`
 - download it from Artifactory [here](https://odbxikk7vo-artifactory.services.clever-cloud.com/gravitee-snapshots/io/gravitee/policy/gravitee-policy-oauth2/1.22.0-feat-jupiterPlans-SNAPSHOT/gravitee-policy-oauth2-1.22.0-feat-jupiterPlans-SNAPSHOT.zip)

### CODE REVIEW

File: src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java [hunk 1]
```diff
@@ -99,10 +97,23 @@ public boolean requireSubscription() {
...
-    public Completable onInvalidSubscription(HttpExecutionContext ctx) {
-        return ctx.interruptWith(
-            new ExecutionFailure(UNAUTHORIZED_401).key(GATEWAY_OAUTH2_ACCESS_DENIED_KEY).message(OAUTH2_ERROR_ACCESS_DENIED)
-        );
+    public Maybe<SecurityToken> extractSecurityToken(HttpExecutionContext ctx) {
+        final OAuth2Resource<?> oauth2Resource = getOauth2Resource(ctx);
+        if (oauth2Resource == null) {
+            log.debug("Skipping security token extraction cause no oauth2 resource configured");
+            return Maybe.empty();
+        }
```
Comment (jhaeyaert) [inline src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java:105]: Does it makes sens to move this into `extractAccessToken` so we can benefit of it when using the policy outside a security policy (ie: in a Flow) ? WDYT ?

File: src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java [hunk 2]
```diff
@@ -228,44 +236,67 @@ private Completable validateOAuth2Payload(
...
+    protected Single<TokenIntrospectionResult> introspectAccessToken(
         HttpExecutionContext ctx,
         String accessToken,
-        CacheResource<?> cacheResource,
         OAuth2Resource<?> oauth2Resource
     ) {
-        return Single
-            .<OAuth2Response>create(emitter -> oauth2Resource.introspect(accessToken, emitter::onSuccess))
+        // find introspection in request context cache
+        TokenIntrospectionCache requestContextCache = getContextTokenIntrospectionCache(ctx);
```
Comment (jhaeyaert) [inline src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java:260]: `tokenIntrospectionCache` ?

File: src/test/java/io/gravitee/policy/oauth2/Oauth2PolicyTest.java [hunk 3]
```diff
@@ -466,46 +471,68 @@ void shouldValidateSubscription() {
     }
 
     @Test
-    void shouldInterruptOnSubscriptionInvalid() {
-        when(ctx.interruptWith(any())).thenReturn(Completable.error(new RuntimeException(MOCK_EXCEPTION)));
+    void extractSecurityToken_shouldReturnEmpty_whenNoOauth2Resource() {
```
Comment (jhaeyaert) [inline src/test/java/io/gravitee/policy/oauth2/Oauth2PolicyTest.java:474]: Can we stay with camelCase 🙏 ?

**Review by jhaeyaert — APPROVED** [review]
Wow, impressive, really !

I let few suggestions but otherwise it looks okay to me.
Here again you've made a huge work on the integration tests, this is fantastic, well done !

File: src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java (hunk 2 above)
Comment (marcambier) [reply src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java:260]: :heavy_check_mark:

File: src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java (hunk 1 above)
Comment (marcambier) [reply src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java:105]: When there is no oauth2 resource configured, the behavior is different :
- during plan selection, extractSecurityToken, returns empty without raising an error
- during policy execution, handleSecurity interrupts with a 401 with a specific error message

So I think we shouldn't mutualize in `extractAccessToken` if we want to keep this distinction.
And also, wouldn't really makes sense to me as extracting the access token from incoming request doesn't need any oauth2 resource at all.

**Review by gcusnieux — APPROVED** [review]

**Pull request merged**